airflow/api_connexion/endpoints/dag_run_endpoint.py

```diff
@@ -271,3 +273,33 @@ def post_dag_run(dag_id, session):
         )
 
     raise AlreadyExists(detail=f"DAGRun with DAG ID: '{dag_id}' and DAGRun ID: '{run_id}' already exists")
+
+
+@security.requires_access(
+    [
+        (permissions.ACTION_CAN_READ, permissions.RESOURCE_DAG),
+        (permissions.ACTION_CAN_EDIT, permissions.RESOURCE_DAG_RUN)
+    ]
+)
+@provide_session
+def post_set_dag_run_state(dag_id: str, session) -> dict:
+    """Set a state of a dag run."""
+    try:
+        post_body = dagrun_schema.load(request.json, session=session, unknown="include")
+    except ValidationError as err:
+        raise BadRequest(detail=str(err))
+    dag_run_id, state = post_body['run_id'], post_body['state']
+    dag_run: Optional[DagRun] = session.query(DagRun).filter(DagRun.dag_id == dag_id, DagRun.run_id == dag_run_id).one_or_none()
```

Comment:
This should happen before loading `post_body`. If a DAG does not exist, the endpoint should always fail with 404.